src/m365/teams/commands/message/message-export.ts

```diff
@@ -0,0 +1,218 @@
+      },
+      {
+        option: '--withAttachments'
+      },
+      {
+        option: '--folderPath [folderPath]'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        if (args.options.userId && !validation.isValidGuid(args.options.userId)) {
+          return `${args.options.userId} is not a valid GUID`;
+        }
+
+        if (args.options.userName && !validation.isValidUserPrincipalName(args.options.userName)) {
+          return `${args.options.userName} is not a valid userPrincipalName`;
+        }
+
+        if (args.options.teamId && !validation.isValidGuid(args.options.teamId)) {
+          return `${args.options.teamId} is not a valid GUID`;
+        }
+
+        if (args.options.fromDateTime && !validation.isValidISODateTime(args.options.fromDateTime)) {
+          return `${args.options.fromDateTime} is not a valid ISO DateTime`;
+        }
+
+        if (args.options.toDateTime && !validation.isValidISODateTime(args.options.toDateTime)) {
+          return `${args.options.toDateTime} is not a valid ISO DateTime`;
+        }
+
+        if (args.options.folderPath && !fs.existsSync(args.options.folderPath)) {
+          return `Path ${args.options.folderPath} does not exist.`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  #initOptionSets(): void {
+    this.optionSets.push(
+      {
+        options: ['userId', 'userName', 'teamId']
+      },
+      {
+        options: ['folderPath'],
+        runsWhen: (args: CommandArgs) => {
+          return args.options.withAttachments !== undefined && args.options.withAttachments;
+        }
+      },
+      {
+        options: ['withAttachments'],
+        runsWhen: (args: CommandArgs) => {
+          return args.options.folderPath !== undefined;
+        }
+      }
+    );
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    let baseUrl = `${this.resource}/v1.0/`;
+
+    if (args.options.userId || args.options.userName) {
+      baseUrl += `users/${args.options.userId || args.options.userName}/chats`;
+    }
+    else {
+      baseUrl += `teams/${args.options.teamId}/channels`;
+    }
+
+    let requestUrl = `${baseUrl}/getAllMessages`;
+
+    const filters = this.getFilters(args.options);
+    if (filters.length > 0) {
+      requestUrl += `?$filter=${filters.join(' and ')}`;
+    }
+
+    try {
+      const res: any[] = await odata.getAllItems(requestUrl);
+      if (args.options.withAttachments) {
+        for (const message of res) {
+          if (message.attachments && message.attachments.length > 0) {
+            for await (const attachment of message.attachments) {
+              const _url = url.parse(attachment['contentUrl']);
+              let siteUrl = _url.protocol + '//' + _url.host!;
+
+              if (_url.path!.split('/')[1] === 'sites' || _url.path!.split('/')[1] === 'teams' || _url.path!.split('/')[1] === 'personal') {
```

Comment:
What's the use of this if-statement?

Reply:
If we do not use it, the siteUrl will always be the root site collection, and an error will be thrown (404) due to the file not being found.

Reply:
No I mean, why do you specifically check for /sites, /teams, /personal?

Reply:
Because, in a Teams chat, it is possible to upload files from a team site, a sharepoint site or a personal onedrive

Reply:
Yes, but what other paths are there? Except maybe custom managed paths, but that was SP on prem I believe.

Reply:
Hi @martinlingstuyl 

This is done to define whether we are taking a file from the root sharepoint site collection or from a site collection, a teams site collection or a OneDrive. 

Hope this clears it up. It's hard to explain by typing 😄

Reply:
> Yes, but what other paths are there? Except maybe custom managed paths, but that was SP on prem I believe.

So, after rereading (I'm really tired, that's why it took 2 times), I think I can explain a little bit more.

So. It is possible to have an attachment on the root site collection of the SharePoint environment. This would make it so that the url becomes for example: https://mathijsdev.sharepoint.com/**Documents**/..., but if we have one on a site collection or onedrive, this would be https://mathijsdev.sharepoint.com/**sites**/demo/Documents/...

If we wouldn't do this check (checking on the /sites/ part, or /teams/ part, we will not be able to retrieve the file using the siteUrl, as only the root site collection part is currently being added to the `siteUrl` variable. 

Does this make sense?

Reply:
Aaahh, now I see why you check for those specific paths 😂